File: week_2/day_09_monotonic/largestRectangleHistogram.js
// File: largestRectangleHistogram.js

function largestRectangleArea(heights) {
    const stack = [];
    heights.push(0); // sentinel
    let maxArea = 0;

    for (let i = 0; i < heights.length; i++) {
        while (stack.length && heights[i] < heights[stack[stack.length - 1]]) {
            const h = heights[stack.pop()];
            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, h * width);
        }
        stack.push(i);
    }
    return maxArea;
}

// INPUT
console.log(largestRectangleArea([2, 1, 5, 6, 2, 3]));
// Output: 10

File: week_2/day_09_monotonic/largest_rectangle_histogram.py
# File: largest_rectangle_histogram.py

def largestRectangleArea(heights):
    stack = []
    max_area = 0
    heights.append(0)  # sentinel to flush stack

    for i in range(len(heights)):
        while stack and heights[i] < heights[stack[-1]]:
            h = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, h * width)
        stack.append(i)

    return max_area


# INPUT
heights = [2, 1, 5, 6, 2, 3]
print(largestRectangleArea(heights))
# Output: 10
